### PR TITLE
Const stylesheet support

### PR DESCRIPTION
**Pull request by WorldSEnder**
Changes the usages of `Vec` in favor of `const` arrays.

### CODE REVIEW

File: packages/stylist-core/src/bow.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use Bow::*;
+
+// Read as "box on write", a copy on write type where the
+// alternative is a Box.
+#[derive(Debug)]
+pub enum Bow<'a, T: 'a + ?Sized> {
```
Comment (WorldSEnder) [inline packages/stylist-core/src/bow.rs:8]: This type is used as an alternative for `Cow` where the `Owned` alternative should be boxed to break cycles, as in `RuleContent`. Should have very similar semantics.

File: packages/stylist-macros/src/output/rule_content.rs [hunk 2]
```diff
@@ -11,19 +11,45 @@ pub enum OutputRuleContent {
 }
 
 impl Reify for OutputRuleContent {
+    #[allow(unreachable_code)]
```
Comment (WorldSEnder) [inline packages/stylist-macros/src/output/rule_content.rs:14]: Oops, will fix.

**Comment by WorldSEnder** [comment]
Preliminary performance comparisons:

Benchmark old: Chrome | runtime
-|-
Parse Simple (10,000,000 iterations): | 495ms
Macro (Literal) Simple (10,000,000 iterations): | 31ms
Macro (Inline) Simple (10,000,000 iterations): | 35ms
Parse Simple, No Cache (100,000 iterations): | 327ms
Parse Complex (1,000,000 iterations): | 899ms
Macro (Literal) Complex (1,000,000 iterations): | 490ms
Macro (Inline) Complex (1,000,000 iterations): | 509ms
Parse Complex, No Cache (100,000 iterations): | 336ms
Cached Lookup (1,000,000 iterations): | 175ms
Cached Lookup, Big Sheet (100,000 iterations): | 364ms
Mounting (2,000 iterations): | 89ms

Benchmark new: Chrome | runtime
-|-
Parse Simple (10,000,000 iterations): | 530ms
Macro (Literal) Simple (10,000,000 iterations): | 31ms
Macro (Inline) Simple (10,000,000 iterations): | 32ms
Parse Simple, No Cache (100,000 iterations): | 302ms
Parse Complex (1,000,000 iterations): | 906ms
Macro (Literal) Complex (1,000,000 iterations): | **134ms**
Macro (Inline) Complex (1,000,000 iterations): | **137ms**
Parse Complex, No Cache (100,000 iterations): | 338ms
Cached Lookup (1,000,000 iterations): | 178ms
Cached Lookup, Big Sheet (100,000 iterations): | 351ms
Mounting (2,000 iterations): | 80ms

Benchmark old: Firefox | runtime
-|-
Parse Simple (10,000,000 iterations):  | 627ms
Macro (Literal) Simple (10,000,000 iterations):  | 67ms
Macro (Inline) Simple (10,000,000 iterations):  | 67ms
Parse Simple, No Cache (100,000 iterations):  | 336ms
Parse Complex (1,000,000 iterations):  | 879ms
Macro (Literal) Complex (1,000,000 iterations):  | 677ms
Macro (Inline) Complex (1,000,000 iterations):  | 737ms
Parse Complex, No Cache (100,000 iterations):  | 341ms
Cached Lookup (1,000,000 iterations):  | 201ms
Cached Lookup, Big Sheet (100,000 iterations):  | 340ms
Mounting (2,000 iterations):  | 515ms

Benchmark new: Firefox | runtime
-|-
Parse Simple (10,000,000 iterations):  | 625ms
Macro (Literal) Simple (10,000,000 iterations):  | 64ms
Macro (Inline) Simple (10,000,000 iterations):  | 75ms
Parse Simple, No Cache (100,000 iterations):  | 325ms
Parse Complex (1,000,000 iterations):  | 877ms
Macro (Literal) Complex (1,000,000 iterations):  | **172ms**
Macro (Inline) Complex (1,000,000 iterations):  | **189ms**
Parse Complex, No Cache (100,000 iterations):  | 340ms
Cached Lookup (1,000,000 iterations):  | 201ms
Cached Lookup, Big Sheet (100,000 iterations):  | 339ms
Mounting (2,000 iterations):  | 527ms

File: packages/stylist-macros/src/output/style_attr.rs [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
+    fragment_coalesce, ContextRecorder, IntoCowVecTokens, OutputCowString, OutputFragment, Reify,
+};
 use itertools::Itertools;
 use proc_macro2::TokenStream;
 use quote::quote;
 use syn::parse::Error as ParseError;
 
 pub struct OutputAttribute {
-    pub key: OutputFragment,
+    pub key: OutputCowString,
```
Comment (futursolo) [inline packages/stylist-macros/src/output/style_attr.rs:10]: This is not directly related to this PR, but rather discovered when I was working on #31.

Is it useful to make the style key substitutable?

File: packages/stylist-macros/src/output/cow_str.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+use super::{ContextRecorder, Reify};
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, quote_spanned, ToTokens};
+use syn::{spanned::Spanned, LitStr};
+
+#[derive(Debug)]
+pub enum OutputCowString {
```
Comment (futursolo) [inline packages/stylist-macros/src/output/cow_str.rs:7]: What's the advantage of stripping this out of `OutputFragment`?

Whilst the code size of procedural macros are not as critical as other parts, I still wish to keep the code footprint low by not having unnecessary abstractions.

**Review by futursolo — CHANGES_REQUESTED** [review]
The performance improvement is quite impressive.

Thank you for the PR.

File: packages/stylist-core/src/ast/rule_content.rs [hunk 5]
```diff
@@ -10,7 +11,7 @@ pub enum RuleContent {
     /// A block
     Block(Block),
     /// A nested rule
-    Rule(Box<Rule>),
+    Rule(Bow<'static, Rule>),
```
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-core/src/ast/rule_content.rs:14]: The moment a nested item isn't `const`, it forces all parents to use a `vec![]` instead of the `const` array, aswell, i.e. they will allocate. Now, given nested rules aren't that common, but they do appear, especially with `@supports`.

File: packages/stylist-macros/src/output/rule_content.rs [hunk 6]
```diff
@@ -11,19 +11,45 @@ pub enum OutputRuleContent {
...
+                let bowed_block = if inner.is_const() {
+                    let const_ident = Ident::new("block", Span::mixed_site());
+                    quote! {
+                        ::stylist::vendor::Bow::Borrowed({
+                            const #const_ident: ::stylist::ast::Rule = #block_tokens;
+                            &#const_ident
+                        })
+                    }
+                } else {
+                    ctx.uses_static(); // Box::new
```
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/rule_content.rs:31]: My reasoning was much simpler: the output quotes an expression that can only be used in `static` context and above. In fact, I should probably move `ctx.uses_nested(&inner);` inside the else block following this logic. While your analysis is of course correct, I think it leads to asking "Why is there no uses_static here?" and more cognitive load than accepting the negligable perf hit of calling `uses_static` unnecessarily.

File: packages/stylist-macros/src/output/context_recorder.rs [hunk 7]
```diff
@@ -14,13 +14,10 @@ enum AllowedUsage {
...
     Static,
-    // TODO: we can probably avoid a few allocations if we track which parts
+    // We can avoid a few allocations if we track which parts
     // of the ast can be constructed statically (with const methods), which is
     // even stronger than constructing it in the global context in a Lazy.
-    // Should you decide to implement this, keep in mind to change Self::MAX
-    // and adjust the generation of cow-vec tokens. Also check the usages of
-    // MaybeStatic::statick if they can be upgraded to Const.
-    // Const,
+    Const,
```
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/context_recorder.rs:20]: Do you mean `Static` and `Dynamic`? At least with the analysis of the superfluous `uses_static` mentioned above, the only place to call `uses_static` is sheet, which doesn't actually care, since there is no `const` option in sheet. Let me explain why conceptually three variants are needed:

- `Const` is ofc connected to `const` in rust, with very specific kinds of expressions that can be used in that context, so far so good. In case an expression that can not be in `const` expressions is used,
- `Static` would be everything that can be sensibly wrapped in a `static Lazy`. Note that I'm a bit unsure if this shouldn't actually be `const Lazy`, the difference is afaik in linkage. Anyway, the important point is that for an expression to qualify as `Static`, it can not mention any extra variables - which wouldn't be available in the closure given to `Lazy::new`.
- `Dynamic` does use local variables from the context and can not be sensibly wrapped in a `Lazy` because of that.

The fact that the expression analysis doesn't currently determine the `AllowedUsage` of an interpolation, apart from special casing string literals doesn't mean we should throw away the distinction between `Const` and `Static`. E.g. if the PR ends up not using `Bow`, then that would imply `Static` instead of `Const` for all parent ast nodes. Ultimately the sheet currently doesn't care, but the intermediate node do.

File: packages/stylist-macros/src/output/context_recorder.rs [hunk 8]
```diff
@@ -1,7 +1,7 @@
 //! This module implements a type abstractly tracking in what kind of expression context
 //! an item appears. This information is leverage to provide improved performance and
 //! static caching of parts of the generated output.
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
 enum AllowedUsage {
```
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/context_recorder.rs:5]: I like the `Kind` part. What about `ExpressionKind`, since it's actually only used to classify those.

File: packages/stylist-macros/src/output/style_attr.rs (hunk 3 above)
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/style_attr.rs:10]: The current declaration of `stylist::ast::StyleAttribute` uses a `Cow<'static, str>`, not a `StringFragment`, hence the change to reflect that reality.

```
pub struct StyleAttribute {
    pub key: Cow<'static, str>,
    pub value: Cow<'static, [StringFragment]>,
}
```

Comment (WorldSEnder) [reply packages/stylist-macros/src/output/style_attr.rs:10]: I do like the idea of making them substitutable, I'm just not sure if we should allow more than one part, e.g. `border-${side}: 5px solid black;` hence I didn't want to include that kind of change here.

File: packages/stylist-macros/src/output/cow_str.rs (hunk 4 above)
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/cow_str.rs:7]: `OutputFragment` returns an expression of type `StringFragment`, `OutputCowString` returns one of type `Cow<'static, str>` from `into_token_stream`. See above why `OutputCowString` is also used in `OutputAttribute`.

File: packages/stylist-macros/src/output/context_recorder.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/context_recorder.rs:20]: `Ord` immensely clarifies the implementation of `uses_nested`, imo, with 3 variants that would otherwise be 6 or up to 9 cases?

[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/context_recorder.rs:20]: Follow up to `Ord`, not using the derive generates worse IR than what I would come up by hand, see e.g. https://godbolt.org/z/8KvGPPhsK Not that it matters in this case, I'll remove the usages of `Ord`.

File: packages/stylist-core/src/ast/rule_content.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (futursolo) [reply packages/stylist-core/src/ast/rule_content.rs:14]: Whilst I suspect the following should work, the compiler actually panics when I change the type to `Cow<'static, Box<T>>`:

```rust
{
    use ::stylist::vendor::once_cell::sync::Lazy;
    static RULE: Lazy<Box<Rule>> = Lazy::new(|| Box::new(#rule_tokens));
    Cow::Borrowed(RULE)
}
```

This is the first time I actually encountered an ICE. :)

```
error: internal compiler error: compiler/rustc_traits/src/normalize_erasing_regions.rs:54:32: could not fully normalize `<std::boxed::Box<ast::rule::Rule> as std::borrow::ToOwned>::Owned`

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1034:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.
```

I also tested `Arc` by replacing all `const` with `Lazy` and achieved a similar performance. But the generated bundle is  slightly bigger. I guess the best course of action for now is to keep `Bow`.

However, `stylist::vendor` is reserved for vendored third party libraries (which should be removed once `std::lazy` is stabilised). Maybe move `Bow` to `stylist::bow` or `stylist::utils`?

File: packages/stylist-macros/src/output/context_recorder.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (futursolo) [reply packages/stylist-macros/src/output/context_recorder.rs:20]: Maybe we can keep `Ord` but give the enum an explicit discriminator so it's less confusing?

File: packages/stylist-macros/src/output/context_recorder.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (futursolo) [reply packages/stylist-macros/src/output/context_recorder.rs:5]: Sounds good to me.

File: packages/stylist-macros/src/output/style_attr.rs (hunk 3 above)
Comment (futursolo) [reply packages/stylist-macros/src/output/style_attr.rs:10]: OK. I guess it's fine to have `OutputCowString` for now.

In the future, maybe the key also needs to be modified to `Cow<'static, [StringFragment]>` to support proper interpolation.

File: packages/stylist-core/src/ast/rule_content.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-core/src/ast/rule_content.rs:14]: I think moving it under `stylist::macros` makes sense, which is afaik how `std::fmt` does it with things like `Arguments`.

File: packages/stylist-macros/src/output/context_recorder.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/context_recorder.rs:35]: instead of `min`, the method is now called `joint_kind` and the usages of `<=` have been replaced by `matches!()`, the performance benefits of deriving `Ord` isn't worth it, imo.

File: packages/stylist-core/src/ast/rule_content.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (futursolo) [reply packages/stylist-core/src/ast/rule_content.rs:14]: I am not very satisfied with the reexport at the root of the documentation.

And after referencing the documentation of `std`, `tokio`, `yew`, `syn` and some other libraries, I think the reasonable action is to remove `macros` module and move macros back to the root of the library.

Which I will do when after the PR list is cleared, so please move `Bow` out of `macros` modules.

File: packages/stylist-macros/src/output/context_recorder.rs [hunk 9]
```diff
@@ -1,37 +1,61 @@
...
+    // ```
+    // style! { color: black; }
+    // //       ------------- everything can be constructed in a const context
+    // ```
+    Const,
+}
+
+impl ExpressionKind {
+    // Compute the jointly allowed usage of two expression that are used together.
+    pub fn joint_kind(l: ExpressionKind, r: ExpressionKind) -> ExpressionKind {
```
Comment (futursolo) [inline packages/stylist-macros/src/output/context_recorder.rs:40]: Adjust function signature to`pub fn join(&self, other: Self) -> Self` like `eq` or `cmp`?

File: packages/stylist-macros/src/output/context_recorder.rs [hunk 10]
```diff
@@ -41,12 +65,24 @@ impl ContextRecorder {
...
+        self.usage = ExpressionKind::joint_kind(self.usage, ExpressionKind::Static);
     }
 
     pub fn is_static(&self) -> bool {
-        self.usage == AllowedUsage::Static
+        matches!(self.usage, ExpressionKind::Static | ExpressionKind::Const)
+    }
+
+    pub fn is_const(&self) -> bool {
+        matches!(self.usage, ExpressionKind::Const)
```
Comment (futursolo) [inline packages/stylist-macros/src/output/context_recorder.rs:86]: `self.usage == ExpressionKind::Const`?

**Review by futursolo — CHANGES_REQUESTED** [review]

File: packages/stylist-core/src/ast/rule_content.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (WorldSEnder) [reply packages/stylist-core/src/ast/rule_content.rs:14]: > I think the reasonable action is to remove macros module 

I think originally I introduced it to have a vehicle to carry the documentation of the syntax so that it doesn't have to be repeated over and over in every single macro. That's also why `tokio::stream` exists. I think it serves that purpose better than it serves as a host for the macro re-exports. Having all the macro exports at top-level would make sense to me though. Lastly, I think the `macros` module serves well as a place to host `vendor` and other things that are used in macro expansion, since it should be `#[cfg(feature = "macros")]`, thus then also turning off the vendored re-exports?

```
stylist
├── sheet!
├── css!
├── global_style!
├── style!
└── macros
    ├── #doc
    ├── #[doc(hidden)] vendor
    │   └── pub use ...
    └── #[doc(hidden)] Bow
```

File: packages/stylist-macros/src/output/context_recorder.rs (hunk 10 above)
Comment (WorldSEnder) [reply packages/stylist-macros/src/output/context_recorder.rs:86]: Mostly so it matches the style used in `is_static`, I think both are mentally the same overhead, close to none.

File: packages/stylist-core/src/ast/rule_content.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (futursolo) [reply packages/stylist-core/src/ast/rule_content.rs:14]: Maybe move Bow to ast?

Bow is currently used by AST regardless of whether feature macros is enabled.

Unless we have feature flags on AST to control which Boxing is used (which I think it's not worth it).

In addition, Bow is also actually visible in the AST(RuleContent::Rule), so I think Bow itself also needs to be visible in the documentation.

As for vendor, I think both stylist::vendor and stylist::macros::vendor are fine.
